lib/grandpa/catch-up.go

```diff
@@ -0,0 +1,281 @@
+	waitingOnResponse *atomic.Value
+	grandpa           *Service
+}
+
+func (c *catchUp) do(to peer.ID, round uint64, setID uint64) error {
+	defer c.waitingOnResponse.Store(false)
+
+	if err := c.sendCatchUpRequest(
+		to, newCatchUpRequest(round, setID),
+	); err != nil {
+		return fmt.Errorf("failed to send catch up request: %w", err)
+	}
+
+	logger.Debugf("successfully sent a catch up request to node %s, for round number %d and set ID %d",
+		to, round, setID)
+
+	c.waitingOnResponse.Store(true)
+
+	timer := time.NewTimer(catchUpResponseTimeout)
+	defer timer.Stop()
+
+	select {
+	case <-c.catchUpResponseCh:
+		return nil
+	case <-timer.C:
+		return errors.New("timeout")
+	}
+}
+
+func (c *catchUp) sendCatchUpRequest(to peer.ID, req *CatchUpRequest) error {
+	if c.bestResponse.Load() != nil {
+		logger.Debug("ignoring neighbour message since we are already processing a catch-up response")
+		return nil
+	}
+
+	// TODO: Clean up all request sent before 5 min / (neighbour message interval)
+	c.lock.Lock()
+	_, ok := c.requestsSent[to]
+	c.lock.Unlock()
+	if ok {
+		logger.Debugf("ignoring neighbour message since we already sent a catch-up request to this peer: %s", to)
+		return nil
+	}
+
+	c.waitingOnResponse.Store(true)
+
+	cm, err := req.ToConsensusMessage()
+	if err != nil {
+		return fmt.Errorf("cannot convert request to consensus message: %w", err)
+	}
+
+	err = c.grandpa.network.SendMessage(to, cm)
+	if err != nil {
+		return fmt.Errorf("cannot send grandpa message: %w", err)
+	}
+
+	c.lock.Lock()
+	c.requestsSent[to] = *req
+	c.lock.Unlock()
+
+	return nil
+}
+
+func (c *catchUp) handleCatchUpResponse(msg *CatchUpResponse) error {
+	if !c.grandpa.authority {
+		return nil
+	}
+
+	logger.Debugf(
+		"processing catch up response with hash %s for round %d and set id %d",
+		msg.Hash, msg.Round, msg.SetID)
+
+	// if we aren't currently expecting a catch up response, return
+	if !c.waitingOnResponse.Load().(bool) {
+		logger.Debug("not currently waiting for a catch up response, ignoring catch up response received")
+		return nil
+	}
+
+	if msg.SetID != c.grandpa.state.setID {
+		return fmt.Errorf("%w: received set id %d but have set id %d in state",
+			ErrSetIDMismatch, msg.SetID, c.grandpa.state.setID)
+	}
+
+	if msg.Round <= c.grandpa.state.round {
+		return fmt.Errorf("%w: received round %d but grandpa round in state is %d",
+			ErrInvalidCatchUpResponseRound, msg.Round, c.grandpa.state.round)
+	}
+
+	if c.bestResponse.Load().(*CatchUpResponse).Round >= msg.Round {
+		logger.Debug("ignoring catch up response, since we are already processing one with a higher round")
+	}
+
+	prevote, err := c.verifyPreVoteJustification(msg)
+	if err != nil {
+		return fmt.Errorf("cannot verify pre vote justification: %w", err)
+	}
+
+	if err = c.verifyPreCommitJustification(msg); err != nil {
+		return fmt.Errorf("cannot verify pre commit justification: %w", err)
+	}
+
+	if msg.Hash.IsEmpty() || msg.Number == 0 {
+		return ErrGHOSTlessCatchUp
+	}
```

Comment:
this could be the second verification step in this function placed right after the `if !c.grandpa.authority`